server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/impl/JamesMailSpoolerTest.java

```diff
@@ -0,0 +1,172 @@
+ ****************************************************************/
+
+package org.apache.james.mailetcontainer.impl;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Duration.ONE_HUNDRED_MILLISECONDS;
+import static org.awaitility.Duration.TEN_SECONDS;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import org.apache.commons.configuration.plist.PropertyListConfiguration;
+import org.apache.james.mailetcontainer.api.MailProcessor;
+import org.apache.james.metrics.api.MetricFactory;
+import org.apache.james.metrics.api.TimeMetric;
+import org.apache.james.queue.api.MailQueue;
+import org.apache.james.queue.api.MailQueueFactory;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.test.FakeMail;
+import org.awaitility.Awaitility;
+import org.awaitility.core.ConditionFactory;
+import org.junit.jupiter.api.Test;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.WorkQueueProcessor;
+
+import javax.mail.MessagingException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+
+class JamesMailSpoolerTest {
+    private static final ConditionFactory CALMLY_AWAIT = Awaitility
+            .with().pollInterval(ONE_HUNDRED_MILLISECONDS)
+            .and().pollDelay(ONE_HUNDRED_MILLISECONDS)
+            .await()
+            .atMost(TEN_SECONDS);
+
+    @Test
+    void thrownExceptionShouldAckTheItem() throws MessagingException {
+        MetricFactory metricFactory = mock(MetricFactory.class);
+        when(metricFactory.timer(JamesMailSpooler.SPOOL_PROCESSING)).thenAnswer(ignored -> mock(TimeMetric.class));
+        JamesMailSpooler spooler = new JamesMailSpooler(metricFactory);
+        MailQueueFactory<?> queueFactory = mock(MailQueueFactory.class);
+
+        WorkQueueProcessor<MockedMailQueueItem> workQueue = WorkQueueProcessor.create("fakeMQ", 1);
+        MockedMailQueueItem item = new MockedMailQueueItem();
+        item.addCallback(isDone -> {
+            if (!isDone) {
+                workQueue.onNext(item);
+            }
+        });
+        MailQueue queue = mock(MailQueue.class);
+        workQueue.onNext(item);
+        when(queue.deQueue()).thenAnswer(any -> workQueue.limitRate(1).filter(MockedMailQueueItem::isNotDone));
+        when(queueFactory.createQueue(MailQueueFactory.SPOOL)).thenAnswer(any -> queue);
+        spooler.setMailQueueFactory(queueFactory);
+
+        MailProcessor mailProcessor = mock(MailProcessor.class);
+        AtomicBoolean hasFailed = new AtomicBoolean(false);
+        doAnswer(ignored -> {
+            if (!hasFailed.getAndSet(true)) {
+                throw new RuntimeException("Arbitrary failure");
+            }
+            return null;
+        }).when(mailProcessor).service(any());
```

Comment:
you don't need a if, you can write it directly with mockito: https://stackoverflow.com/questions/11785498/simulate-first-call-fails-second-call-succeeds/41272746#41272746